lib/Chat/Notifier.php

```diff
@@ -170,18 +171,19 @@ private function addMentionAllToList(Room $chat, array $list): array {
 			return $usersToNotify;
 		}
 
-		$chatParticipants = $this->participantService->getActorsByType($chat, Attendee::ACTOR_USERS);
-		foreach ($chatParticipants as $participant) {
-			$alreadyAddedToNotify = array_filter($list, static function ($user) use ($participant): bool {
-				return $user['id'] === $participant->getActorId();
+		$attendees = $this->participantService->getActorsByType($chat, Attendee::ACTOR_USERS);
+		foreach ($attendees as $attendee) {
+			$alreadyAddedToNotify = array_filter($list, static function ($user) use ($attendee): bool {
+				return $user['id'] === $attendee->getActorId();
 			});
 			if (!empty($alreadyAddedToNotify)) {
 				continue;
 			}
 
 			$usersToNotify[] = [
-				'id' => $participant->getActorId(),
-				'type' => $participant->getActorType()
+				'id' => $attendee->getActorId(),
+				'type' => $attendee->getActorType(),
+				'attendee' => $attendee,
```

Comment:
Until now all the element in the `usersToNotify` arrays generated by the different methods have just `id` and `type`, so it was safe to remove already notified users by checking the full item. Although currently it should make no difference given the order in which the methods are called it would be better to change https://github.com/nextcloud/spreed/blob/e907617e93cac5a56c5ac9ac78a0de136912319b/lib/Chat/Notifier.php#L149 so it explicitly compares the `id` and `type` fields instead of the full item.

Besides that, the different `@psalm-return array<int, array{id: string, type: string}>` docs should be updated to reflect the optional `attendee` parameter.

Reply:
Will be more complex to refactor all methods and add attendee, do this maybe will lost the performance to get all data of attendee from id and type.

I think that the best is fix the psalm and unit tests to cover the scenario when the attendee is optional.

Reply:
Will be more complex to refactor all methods and add attendee, do this maybe will lost the performance to get all data of attendee from id and type.

I think that the best is fix the psalm and unit tests to cover the scenario when the attendee is optional.

Reply:
Yeah, was only a quick commit before I started into the weekend.
As we always get the object in the end, we can also get them ahead of time and then only deal with attendee objects in all those methods?

Reply:
Enjoy the weekend. I believe that now the tests will pass.
We got a excellent reduction with your changes. Excellent work!

Reply:
Done! All tests are green.